In [1]:
import pandas as pd
import numpy as np
data=pd.read_csv("C:/Users/DANIYA/Desktop/PROJECT/og.csv")
df=pd.DataFrame(data)
df.head()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 175341 entries, 0 to 175340
Data columns (total 45 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   id                 175341 non-null  int64  
 1   dur                175341 non-null  float64
 2   proto              175341 non-null  object 
 3   service            175341 non-null  object 
 4   state              175341 non-null  object 
 5   spkts              175341 non-null  int64  
 6   dpkts              175341 non-null  int64  
 7   sbytes             175341 non-null  int64  
 8   dbytes             175341 non-null  int64  
 9   rate               175341 non-null  float64
 10  sttl               175341 non-null  int64  
 11  dttl               175341 non-null  int64  
 12  sload              175341 non-null  float64
 13  dload              175341 non-null  float64
 14  sloss              175341 non-null  int64  
 15  dloss              175341 non-null  int64  
 16  si

# decision tree og

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

df = df.drop(columns=['id', 'attack_cat'])
X = df.drop(columns=['label'])  # Features
y = df['label']  # Target variable
X = pd.get_dummies(X)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
dt_classifier = DecisionTreeClassifier(random_state=42)
dt_classifier.fit(X_train, y_train)
y_pred = dt_classifier.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy}')
feature_importances = dt_classifier.feature_importances_
importance_df = pd.DataFrame({'Feature': X.columns, 'Importance': feature_importances})
importance_df = importance_df.sort_values(by='Importance', ascending=False)
print("\nTop 10 Most Important Features:")
print(importance_df.head(10))


Accuracy: 0.9482734038609598

Top 10 Most Important Features:
           Feature  Importance
6             sttl    0.676624
37      ct_srv_dst    0.063202
3           sbytes    0.042595
23           smean    0.031890
27      ct_srv_src    0.014017
20          tcprtt    0.010605
12          sinpkt    0.010279
17           stcpb    0.009931
32  ct_dst_src_ltm    0.009885
18           dtcpb    0.009591


In [3]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Your DataFrame and data preprocessing code here

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
dt_classifier = DecisionTreeClassifier(random_state=42)
dt_classifier.fit(X_train, y_train)
y_pred_train = dt_classifier.predict(X_train)  # Predict on the training data
accuracy_train = accuracy_score(y_train, y_pred_train)  # Calculate training accuracy
print(f'Training Accuracy: {accuracy_train}')


Training Accuracy: 0.9983460704916163


# evaluation of the above accuracy with the use of precision , recall , f1-score

In [4]:
from sklearn.metrics import precision_score, recall_score, f1_score
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
print(f'Precision: {precision}')
print(f'Recall: {recall}')
print(f'F1-score: {f1}')


Precision: 0.9611241022214799
Recall: 0.9630543933054393
F1-score: 0.9620882795519144


In [5]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score


dt_classifier = DecisionTreeClassifier(random_state=42)
dt_classifier.fit(X_train, y_train)

feature_importances = dt_classifier.feature_importances_

importance_df = pd.DataFrame({'Feature': X.columns, 'Importance': feature_importances})

importance_df = importance_df.sort_values(by='Importance', ascending=False)

print("\nTop 10 Most Important Features:")
print(importance_df.head(10))

selected_features = importance_df['Feature'][:10]  
print("\nSelected Features:")
print(selected_features)




Top 10 Most Important Features:
           Feature  Importance
6             sttl    0.676624
37      ct_srv_dst    0.063202
3           sbytes    0.042595
23           smean    0.031890
27      ct_srv_src    0.014017
20          tcprtt    0.010605
12          sinpkt    0.010279
17           stcpb    0.009931
32  ct_dst_src_ltm    0.009885
18           dtcpb    0.009591

Selected Features:
6               sttl
37        ct_srv_dst
3             sbytes
23             smean
27        ct_srv_src
20            tcprtt
12            sinpkt
17             stcpb
32    ct_dst_src_ltm
18             dtcpb
Name: Feature, dtype: object


In [6]:
feature_importances 

array([8.68260209e-03, 3.60568361e-04, 1.45837631e-03, 4.25947927e-02,
       7.63740536e-03, 4.90471433e-03, 6.76624209e-01, 0.00000000e+00,
       9.51719494e-03, 5.62567302e-03, 7.01444818e-03, 2.16724435e-03,
       1.02788673e-02, 6.08159037e-03, 6.02087208e-03, 5.92692859e-03,
       0.00000000e+00, 9.93054638e-03, 9.59092405e-03, 0.00000000e+00,
       1.06049796e-02, 8.64101094e-03, 7.68001562e-03, 3.18901440e-02,
       7.26302428e-03, 5.04615253e-04, 6.18451961e-04, 1.40165661e-02,
       1.89776217e-05, 5.02296469e-03, 3.69337406e-03, 2.42029751e-03,
       9.88460466e-03, 0.00000000e+00, 0.00000000e+00, 5.49552478e-04,
       6.19596421e-03, 6.32020122e-02, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 3.31905099e-03, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
      

# removing dmean n smean

from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

data=pd.read_csv("C:/Users/DANIYA/Desktop/PROJECT/og.csv")
df=pd.DataFrame(data)
df = df.drop(columns=['id', 'attack_cat','smean','dmean'])

selector = SelectKBest(f_classif, k=10)
X_train_top10 = selector.fit_transform(X_train, y_train)
X_test_top10 = selector.transform(X_test)
dt_classifier = DecisionTreeClassifier(random_state=42)
dt_classifier.fit(X_train_top10, y_train)
y_pred_top10 = dt_classifier.predict(X_test_top10)
accuracy_top10 = accuracy_score(y_test, y_pred_top10)
print(f'Accuracy with Top 10 Features: {accuracy_top10}')



selected_feature_indices = selector.get_support(indices=True)
selected_feature_names = X.columns[selected_feature_indices]
print("Selected Features:")
for feature in selected_feature_names:
    print(feature)

# iteration using max depth

In [7]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

max_depths = [None, 20, 30, 40]
results = {}

for max_depth in max_depths:
    dt_classifier = DecisionTreeClassifier(max_depth=max_depth, random_state=42)
    dt_classifier.fit(X_train, y_train)
    y_pred = dt_classifier.predict(X_test)

    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)

    results[max_depth] = {
        'accuracy': accuracy,
        'precision': precision,
        'recall': recall,
        'f1_score': f1
    }

print("Performance Metrics for Different Max Depths:")
for max_depth, metrics in results.items():
    print(f"Max Depth {max_depth}:")
    print(f"Accuracy = {metrics['accuracy']:.4f}")
    print(f"Precision = {metrics['precision']:.4f}")
    print(f"Recall = {metrics['recall']:.4f}")
    print(f"F1 Score = {metrics['f1_score']:.4f}")
    print()


Performance Metrics for Different Max Depths:
Max Depth None:
Accuracy = 0.9483
Precision = 0.9611
Recall = 0.9631
F1 Score = 0.9621

Max Depth 20:
Accuracy = 0.9497
Precision = 0.9623
Recall = 0.9639
F1 Score = 0.9631

Max Depth 30:
Accuracy = 0.9491
Precision = 0.9617
Recall = 0.9636
F1 Score = 0.9627

Max Depth 40:
Accuracy = 0.9486
Precision = 0.9613
Recall = 0.9634
F1 Score = 0.9623



The accuracy scores for different max depths are displayed in the output.
Generally, you observe that accuracy is relatively high for all max_depth values, ranging from approximately 94.82% to 94.97%.
These accuracy scores provide insights into how different values of the max_depth hyperparameter affect the model's performance. In this case, it seems that the model performs well across a range of max_depth values, with no substantial improvement observed by increasing the depth beyond 20.